src/Flysystem/FlysystemAssetStore.php

```diff
@@ -913,24 +954,94 @@ public static function flush()
 
     public function getResponseFor($asset)
     {
-        // Check if file exists
-        $filesystem = $this->getFilesystemFor($asset);
-        if (!$filesystem) {
-            return $this->createMissingResponse();
+
+        $public = $this->getPublicFilesystem();
+        $protected = $this->getProtectedFilesystem();
+
+        // If the file exists on the public store, we just straight return it.
+        if ($public->has($asset)) {
+            return $this->createResponseFor($public, $asset);
         }
 
-        // Block directory access
-        if ($filesystem->get($asset) instanceof Directory) {
-            return $this->createDeniedResponse();
+        // If the file exists in the protected store and the user has been explicitely granted access to it
+        if ($protected->has($asset) && $this->isGranted($asset)) {
+            return $this->createResponseFor($protected, $asset);
+            // Let's not deny if the file is in the protected store, but is not granted.
+            // We might be able to redirect to a live version.
+        }
+
+        // If we found a URL to redirect to
+        if ($redirectUrl = $this->searchForEquivalentFileID($asset)) {
+            if ($redirectUrl != $asset && $public->has($redirectUrl)) {
+                return $this->createRedirectResponse($redirectUrl);
+            } else {
+                // Something weird is going on e.g. a publish file without a physical file
+                return $this->createMissingResponse();
+            }
         }
 
         // Deny if file is protected and denied
-        if ($filesystem === $this->getProtectedFilesystem() && !$this->isGranted($asset)) {
+        if ($protected->has($asset)) {
             return $this->createDeniedResponse();
         }
 
-        // Serve up file response
-        return $this->createResponseFor($filesystem, $asset);
+        // We've looked everywhere and couldn't find a file
+        return $this->createMissingResponse();
+    }
+
+    /**
+     * @return string
+     */
+    private function searchForEquivalentFileID($asset)
```

Comment:
Made this private because I don't want to introduce new APIs in a patch release and limit my option for the SS44 fix.